lua/starfall/sflib.lua

```diff
@@ -19,45 +23,63 @@ hook.Add("InitPostEntity","SF_SanitizeTypeMetatables",function()
 					local myMetaFunc = myMeta and myMeta[k]
 					if myMetaFunc then
 						meta[k] = function(...)
-							if SF.runningOps then return myMetaFunc(...) else return v(...) end
+							if SF.runningOps and getfenv(2) ~= _G then
+								return myMetaFunc(...)
+							else
+								return v(...)
+							end
 						end
 					else
 						meta[k] = function(...)
-							if not SF.runningOps then return v(...) end
+							if not SF.runningOps or getfenv(2) == _G then
+								return v(...)
+							end
 						end
 					end
 				elseif istable(v) and k=="__index" then
 					local myMetaFunc = myMeta and myMeta[k]
 					if myMetaFunc then
 						meta[k] = function(t,k)
-							if SF.runningOps then return myMetaFunc(t,k) else return rawget(t,k) end
+							if SF.runningOps and getfenv(2) ~= _G then
+								return myMetaFunc(t,k)
+							else
+								return rawget(t,k)
+							end
 						end
 					else
 						meta[k] = function(t,k)
-							if not SF.runningOps then return rawget(t,k) end
+							if not SF.runningOps or getfenv(2) == _G then
+								return rawget(t,k)
+							end
 						end
 					end
 				end
 			end
 		end
 	end
 
+
 	sanitizeTypeMeta(nil)
 	sanitizeTypeMeta(true)
 	sanitizeTypeMeta(0)
 	sanitizeTypeMeta(function() end)
 	sanitizeTypeMeta(coroutine.create(function() end))
 
-	local string_library = SF.SafeStringLib
 	local function sf_string_index(self, key)
-		local val = string_library[key]
-		if (val) then
-			return val
-		elseif (tonumber(key)) then
-			return self:sub(key, key)
-		else
-			SF.Throw("attempt to index a string value with bad key ('" .. tostring(key) .. "' is not part of the string library)", 2)
+		local instance_string = SF.runningOps.Libraries.string
+		local value = rawget(instance_string, key)
+		if value then
+			return value
+		end
+		local meta = debug.getmetatable(instance_string)
```

Comment:
I think debug.getmetatable can be changed to dgetmeta. The changes look better than before. I will test out the performance as well.

Reply:
Good catch.